### DISCUSSION TITLE
Adding additional token provider to identity service from module

### DISCUSSION THREAD

**Discussion by Dnawrkshp** [oqtane/oqtane.framework#903] [Q&A]
Hi I'm trying to add a passwordless authentication to my company's Oqtane module. I'm following a tutorial [here](https://andrewlock.net/implementing-custom-token-providers-for-passwordless-authentication-in-asp-net-core-identity/) that creates a token provider and adds it to an identity service.

```cs
public class Startup  
{
    public void ConfigureServices(IServiceCollection services)
    {
        services.AddIdentity<IdentityUser, IdentityRole>()
            .AddEntityFrameworkStores<IdentityDbContext>() 
            .AddDefaultTokenProviders()
            .AddPasswordlessLoginTotpTokenProvider(); // Add the custom token provider
    }
}
```

Since we're developing a module we don't have the ability to modify Oqtane's startup. Implementing IServerStartup and adding another identity service seems incorrect. If I were able to modify Oqtane's startup.cs, I could just add to the end of the chain like so:

```cs
// From oqtane.framework Startup.cs
services.AddIdentityCore<IdentityUser>(options => { })
  .AddEntityFrameworkStores<TenantDBContext>()
  .AddSignInManager()
  .AddDefaultTokenProviders()
  .AddPasswordlessLoginTotpTokenProvider();
```

Is there any way for me to accomplish something like this within a module?

Thanks!

**Comment by hishamco** [comment]
The simplest approach is to replace the registered services

**Comment by hishamco** [comment]
@sbwalker IMHO we need to make such services easy to extend and replaceable instead of replacing DI services. In the above scenario we could introduce `ITokenProvider` which have the default implement, but whenever need a custom implementation say `PasswordlessTokenProvider` just registering the new implementation in DI is enough to let Oqtane take care about the rest. I can provide a PR if this make sense

**Reply by sbwalker** [reply]
Sounds feasible - the only problem would be if multiple modules try to provide an ITokenProvider implementation... as this is a situation where the system could only deal with 1 active alternative.

**Reply by hishamco** [reply]
The DI should track `ITokenProvider` not `IEnumerable<ITokenProvider>`, but if we do so the last registered provide should when, this is similar to other services

**Reply by hishamco** [reply]
If I recall DI services have `AddXXX` and `TryAddXXX` versions which is make such situation possible to have only one active implementation, I can propose a draft PR

**Reply by sbwalker** [reply]
I have another suggestion... rather than introduce more interfaces ( which will impose a performance hit on startup ) why not modify the core framework to use the existing IServerStartup interface to dynamically register services which can be replaced? This is the whole purpose of IServerStartup and if we expect modules to use it then it makes sense that the core framework should use it as well ( as this provides a functional implementation ). Basically we would create a number of granular implementations of the IServerStartup interface for various services that are currently hardcoded in main Startup class, and include feature flags to enable/disable them. Then if a third party developer wanted to register their own Identity service configuration they could implement IServerStartup and disable the default Identity service registration.

**Reply by hishamco** [reply]
@sbwalker I started on a proof of concept of what you are suggested, the only issue I faces is the order of the registered coz you knew very well that the order is `Startup` is very important, that's why I suggested to create an interface per core service with that developers can extend those by implementing them, and the system only knows how to inject them at the right place

I will push a draft PR soon ...

**Reply by hishamco** [reply]
The PR #952 is created

**Reply by hishamco** [reply]
Seems prioritizing startup interfaces introduces a breaking change, so I think differently and I come up with another approach that I need to test before I share it with you ;)

**Reply by sbwalker** [reply]
@hishamco I am currently documenting the pros/cons of different approaches and will create a new Discussion thread. You will be able to add a description of your ideas as well.

**Comment by sbwalker** [comment]
Moving discussion to #956 as this is not specific to authentication and touches on a more general problem related to Startup Extensibility.
